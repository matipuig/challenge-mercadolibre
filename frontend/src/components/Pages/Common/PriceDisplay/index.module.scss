@use '~/scss/variables/fontSizes' as fontSizes;

.container {
  font-size: inherit;
  span {
    font-size: inherit;
  }
}
