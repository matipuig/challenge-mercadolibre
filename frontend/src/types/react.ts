/**
 * Contains the different types necessary to use React with typescript.
 */
import React, { PropsWithChildren, ReactElement as Element } from 'react';

export type ComponentChildren = React.ReactNode;
export type ReactElement = Element;
export type FunctionalComponent<P = unknown> = (props: P) => ReactElement | null;
export type FunctionalComponentWithChildren<P = unknown> = FunctionalComponent<
  PropsWithChildren<P>
>;
