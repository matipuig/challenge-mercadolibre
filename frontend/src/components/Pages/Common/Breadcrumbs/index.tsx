/**
 * Contains the breadcrumbs component.
 */
import { Fragment, ReactElement } from 'react';

import { t } from 'i18next';
import Link from 'next/link';

import { CONSTANTS } from '~/constants';
import { getAvailableI18nTexts, i18nReplace } from '~/i18n';
import { ItemCategory } from '~/types/services/backend';
import { getURLFriendlyString } from '~/utils/queryParams';

import styles from './index.module.scss';

interface BreadcrumbProps {
  categories: ItemCategory[];
}

const { ROUTES } = CONSTANTS;
const { QUERY_PARAMS } = ROUTES;
const texts = getAvailableI18nTexts();
const { ariaLabel } = texts.components.common.breadcrumbs;

const getDestinationRoute = (category: ItemCategory): string => {
  const { id, name } = category;
  const queryParams = new URLSearchParams();
  const friendlyCategoryName = getURLFriendlyString(name);
  queryParams.set('category', id);
  return `${ROUTES.ITEMS}?${queryParams}&${QUERY_PARAMS.DESCRIPTION}=${friendlyCategoryName}`;
};

export const Breadcrumbs = ({ categories }: BreadcrumbProps): ReactElement => {
  const lastCategoryIndex = categories.length - 1;
  return (
    <nav className={styles.breadcrumbsContainer} data-testid="breadcrumbs">
      {categories.map((category, i) => (
        <Fragment key={category.id}>
          <Link href={getDestinationRoute(category)}>
            <a
              className={styles.breadcrumb}
              data-testid="breadcrumbLink"
              aria-label={i18nReplace(t(ariaLabel), category.name)}
            >
              {category.name}
            </a>
          </Link>
          {i < lastCategoryIndex && (
            <div className={styles.separator} aria-hidden>
              &gt;
            </div>
          )}
        </Fragment>
      ))}
    </nav>
  );
};

export default Breadcrumbs;
