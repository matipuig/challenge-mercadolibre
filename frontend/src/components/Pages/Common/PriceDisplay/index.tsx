/**
 * Contains the no results screen.
 */
import { ReactElement } from 'react';

import { useTranslation } from 'react-i18next';

import { getAvailableI18nTexts, i18nReplace } from '~/i18n';
import { Price } from '~/types/services/backend';
import { makeHumanFriendly } from '~/utils/conversors';

import styles from './index.module.scss';

interface PriceDisplayProps {
  price: Price;
  withDecimals?: boolean;
}

const texts = getAvailableI18nTexts();
const { withPesos } = texts.components.common.priceDisplay;

const getDecimals = (decimals: number): string => {
  const strDecimals = decimals.toString();
  return strDecimals.length < 2 ? `0${strDecimals}` : strDecimals;
};

export const PriceDisplay = ({ price, withDecimals }: PriceDisplayProps): ReactElement => {
  const { t } = useTranslation();
  const { amount, decimals, currency } = price;
  const friendlyAmount = makeHumanFriendly(amount);
  const isPesos = currency === 'ARS';
  const currencyDisplay = isPesos ? '$' : currency;
  const ariaLabel = isPesos ? i18nReplace(t(withPesos), amount) : `${amount} ${currency}`;
  const finalDecimals = getDecimals(decimals);
  return (
    <div className={styles.container} aria-label={ariaLabel} data-testid="priceDisplay">
      <span className={styles.amount}>
        {currencyDisplay} {friendlyAmount}
      </span>
      {withDecimals && (
        <span aria-hidden className={styles.decimals}>
          <sup>{finalDecimals}</sup>
        </span>
      )}
    </div>
  );
};

export default PriceDisplay;
