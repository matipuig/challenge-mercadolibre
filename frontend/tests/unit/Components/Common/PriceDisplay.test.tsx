/**
 * Contains the unit tests for PriceDisplay.
 */
import { cleanup, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { mockedPrices } from '../../__mocks__/prices';
import { PriceDisplay } from '../../../../src/components/Pages/Common/PriceDisplay';

describe('Common price display component', () => {
  afterEach(() => {
    cleanup();
  });

  it('renders.', () => {
    const priceDisplay = render(<PriceDisplay price={mockedPrices[0]} />);
    expect(priceDisplay.baseElement).toBeInTheDocument();
  });

  it('with pesos matches snapshot.', () => {
    const priceDisplay = render(<PriceDisplay price={mockedPrices[0]} />);
    expect(priceDisplay).toMatchSnapshot();
  });

  it('with other currency matches snapshot.', () => {
    const priceDisplay = render(<PriceDisplay price={mockedPrices[1]} />);
    expect(priceDisplay).toMatchSnapshot();
  });
});
