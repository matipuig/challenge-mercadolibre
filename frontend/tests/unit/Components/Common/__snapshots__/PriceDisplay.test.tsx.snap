// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common price display component with other currency matches snapshot. 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-label="100 US"
        data-testid="priceDisplay"
      >
        <span>
          US
           
          100
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-label="100 US"
      data-testid="priceDisplay"
    >
      <span>
        US
         
        100
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Common price display component with pesos matches snapshot. 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-label="500 pesos"
        data-testid="priceDisplay"
      >
        <span>
          $
           
          500
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-label="500 pesos"
      data-testid="priceDisplay"
    >
      <span>
        $
         
        500
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
